sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -0,0 +1,353 @@
+      physicalPlanDescription, sparkPlanInfo, time) = event
+
+    def toStoredNodes(nodes: Seq[SparkPlanGraphNode]): Seq[SparkPlanGraphNodeWrapper] = {
+      nodes.map {
+        case cluster: SparkPlanGraphCluster =>
+          val storedCluster = new SparkPlanGraphClusterWrapper(
+            cluster.id,
+            cluster.name,
+            cluster.desc,
+            toStoredNodes(cluster.nodes),
+            cluster.metrics)
+          new SparkPlanGraphNodeWrapper(null, storedCluster)
+
+        case node =>
+          new SparkPlanGraphNodeWrapper(node, null)
+      }
+    }
+
+    val planGraph = SparkPlanGraph(sparkPlanInfo)
+    val sqlPlanMetrics = planGraph.allNodes.flatMap { node =>
+      node.metrics.map { metric => (metric.accumulatorId, metric) }
+    }.toMap.values.toList
+
+    val graphToStore = new SparkPlanGraphWrapper(
+      executionId,
+      toStoredNodes(planGraph.nodes),
+      planGraph.edges)
+    kvstore.write(graphToStore)
+
+    val exec = getOrCreateExecution(executionId)
+    exec.description = description
+    exec.details = details
+    exec.physicalPlanDescription = physicalPlanDescription
+    exec.metrics = sqlPlanMetrics
+    exec.submissionTime = time
+    update(exec)
+  }
+
+  private def onExecutionEnd(event: SparkListenerSQLExecutionEnd): Unit = {
+    val SparkListenerSQLExecutionEnd(executionId, time) = event
+    liveExecutions.get(executionId).foreach { exec =>
+      synchronized {
+        exec.metricsValues = aggregateMetrics(exec)
+
+        // Remove stale LiveStageMetrics objects for stages that are not active anymore.
+        val activeStages = liveExecutions.values.flatMap { other =>
+          if (other != exec) other.stages else Nil
+        }.toSet
+        stageMetrics.retain { case (id, _) => activeStages.contains(id) }
+
+        exec.completionTime = Some(new Date(time))
+        exec.endEvents += 1
+
+        update(exec)
+      }
+    }
+  }
+
+  private def onDriverAccumUpdates(event: SparkListenerDriverAccumUpdates): Unit = {
+    val SparkListenerDriverAccumUpdates(executionId, accumUpdates) = event
+    liveExecutions.get(executionId).foreach { exec =>
+      exec.driverAccumUpdates = accumUpdates.toMap
+      update(exec)
+    }
+  }
+
+  override def onOtherEvent(event: SparkListenerEvent): Unit = event match {
+    case e: SparkListenerSQLExecutionStart => onExecutionStart(e)
+    case e: SparkListenerSQLExecutionEnd => onExecutionEnd(e)
+    case e: SparkListenerDriverAccumUpdates => onDriverAccumUpdates(e)
+    case _ => // Ignore
+  }
+
+  private def getOrCreateExecution(executionId: Long): LiveExecutionData = {
+    liveExecutions.getOrElseUpdate(executionId, new LiveExecutionData(executionId))
+  }
+
+  private def update(exec: LiveExecutionData): Unit = {
+    val now = System.nanoTime()
+    if (exec.endEvents >= exec.jobs.size + 1) {
+      liveExecutions.remove(exec.executionId)
+      exec.write(kvstore, now)
```

Comment:
if a job was killed mid execution, so the event log didn't contain the end events, you wouldn't get to see any info for that execution, would you?  don't you need some final flush?

Reply:
When you kill a job, isn't an end event generated?

This is trying to mimic the behavior from the old listener:

```
      jobEnd.jobResult match {
        case JobSucceeded => executionUIData.jobs(jobId) = JobExecutionStatus.SUCCEEDED
        case JobFailed(_) => executionUIData.jobs(jobId) = JobExecutionStatus.FAILED
      }
      if (executionUIData.completionTime.nonEmpty && !executionUIData.hasRunningJobs) {
        // We are the last job of this execution, so mark the execution as finished. Note that
        // `onExecutionEnd` also does this, but currently that can be called before `onJobEnd`
        // since these are called on different threads.
        markExecutionFinished(executionId)
      }
```

So in that case, if there's no end event for the job, the execution will not be marker as finished either.

Reply:
sorry, I meant if the _application_ is killed, so the event log just ends abruptly.  I think the old history server code would still show you the updated metrics for all the tasks that had completed.  But seems like after this change, the history server won't show anything for any jobs which hadn't completed.

Reply:
That's true; I've actually fixed that in the M6 code:
https://github.com/vanzin/spark/pull/51/files#diff-a74d84702d8d47d5269e96740a55a3caR56

It's not very easy to fix here without writing throw-away code to propagate some "close()" call to this listener.